src/raftstore/store/snap.rs

```diff
@@ -61,223 +108,533 @@ impl Display for SnapKey {
+                snap: &DbSnapshot,
+                region: &Region,
+                snap_data: &mut RaftSnapshotData)
+                -> raft::Result<()>;
+}
 
-/// A structure represents the snapshot file.
-///
-/// All changes to the file will be written to `tmp_file` first, and use
-/// `save` method to make them persistent. When saving a crc32 checksum
-/// will be appended to the file end automatically.
-pub struct SnapFile {
-    file: PathBuf,
-    digest: Digest,
-    size_track: Arc<RwLock<u64>>,
-    // File is the file obj represent the tmpfile, string is the actual path to
-    // tmpfile.
-    tmp_file: Option<(File, String)>,
+pub trait SnapshotCommon {
+    fn path(&self) -> &str;
+    fn exists(&self) -> bool;
+    fn delete(&self);
 }
 
-impl SnapFile {
-    fn new<T: Into<PathBuf>>(snap_dir: T,
-                             size_track: Arc<RwLock<u64>>,
-                             is_sending: bool,
-                             key: &SnapKey)
-                             -> io::Result<SnapFile> {
-        let mut file_path = snap_dir.into();
-        if !file_path.exists() {
-            try!(fs::create_dir_all(file_path.as_path()));
-        }
-        let prefix = if is_sending {
-            SNAP_GEN_PREFIX
-        } else {
-            SNAP_REV_PREFIX
-        };
-        let file_name = format!("{}_{}.snap", prefix, key);
-        file_path.push(&file_name);
-
-        let mut f = SnapFile {
-            file: file_path,
-            digest: Digest::new(crc32::IEEE),
-            size_track: size_track,
-            tmp_file: None,
-        };
-        try!(f.init());
-        Ok(f)
-    }
-
-    pub fn init(&mut self) -> io::Result<()> {
-        if self.exists() || self.tmp_file.is_some() {
-            return Ok(());
-        }
+/// `Snapshot` is a trait for snapshot.
+/// It's used in these scenarios:
+///   1. read local snapshot and then replicate it to remote raftstores
+///   2. receive snapshot from remote raftstore and write it to local storage
+///   3. snapshot gc
+pub trait Snapshot: SnapshotCommon + Read + Write + Send {
+    fn meta(&self) -> io::Result<Metadata>;
+    fn total_size(&self) -> io::Result<u64>;
+    fn save(&mut self) -> io::Result<()>;
+}
 
-        let tmp_path = format!("{}{}", self.path().display(), TMP_FILE_SUFFIX);
-        let tmp_f = try!(OpenOptions::new().write(true).create_new(true).open(&tmp_path));
-        self.tmp_file = Some((tmp_f, tmp_path));
-        Ok(())
-    }
+pub struct ApplyOptions<'a> {
+    pub db: Arc<DB>,
+    pub region: &'a Region,
+    pub abort: Arc<AtomicUsize>,
+    pub write_batch_size: usize,
+    pub snapshot_size: &'a mut usize,
+    pub snapshot_kv_count: &'a mut usize,
+}
 
-    pub fn meta(&self) -> io::Result<Metadata> {
-        self.file.metadata()
-    }
+/// `SnapshotApplier` is a trait that applies the snapshot which is received from
+/// remote raftstore to local `RocksDB`.
+pub trait SnapshotApplier: SnapshotCommon {
```

Comment:
> Then it remove necessariness of the Applier trait

I think it is fine because of your following logic. I thought before that the applier interface should be `apply(options, snapshot)`, but here you just use `apply(options)`, no any snapshot information, so I think it is better to move `apply` to the Snapshot trait directly.

> we could remove Generator trait. It's fine for v1 impl, since there only one Snap. But it combines too much code for v2 impl IMO.
Why Generator has too much code for v2?